File: src/ovi_voice_assistant/tts/kokoro_tts.py
"""Streaming text-to-speech using Kokoro 82M (ONNX int8)."""

import logging
import os
import urllib.request
from collections.abc import Iterable
from pathlib import Path

import numpy as np
import onnxruntime as ort

from ovi_voice_assistant.config import CACHE_DIR, Settings
from ovi_voice_assistant.tts import resample
from ovi_voice_assistant.tts.tts import TTS

logger = logging.getLogger(__name__)

MODELS_DIR = CACHE_DIR / "kokoro"

_MODEL_URL = (
    "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX"
    "/resolve/main/onnx/model_uint8.onnx"
)
_VOICES_URL = (
    "https://github.com/thewh1teagle/kokoro-onnx"
    "/releases/download/model-files-v1.0/voices-v1.0.bin"
)

_NATIVE_RATE = 24000


class KokoroTTS(TTS):
    """Text-to-speech using Kokoro 82M (ONNX int8 quantized).

    Uses kokoro-onnx for phonemization, tokenization, and voice loading.
    Runs ONNX inference directly to handle input type requirements of
    quantized models (kokoro-onnx passes wrong dtypes for some inputs).
    """

    def __init__(self, settings: Settings, sample_rate: int = 16000) -> None:
        self._settings = settings
        self._kokoro = None
        self._input_dtypes: dict[str, str] = {}
        self._target_rate = sample_rate
        self.sample_rate: int = sample_rate
        self.sample_width: int = 2  # 16-bit
        self.channels: int = 1  # mono

    def load(self) -> None:
        from kokoro_onnx import Kokoro

        model_path = self._ensure_file("model_uint8.onnx", _MODEL_URL)
        voices_path = self._ensure_file("voices-v1.0.bin", _VOICES_URL)

        logger.info("Loading Kokoro TTS model")

        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = (
            ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
        )
        sess_options.execution_mode = ort.ExecutionMode.ORT_PARALLEL
        threads = os.cpu_count() or 4
        sess_options.intra_op_num_threads = threads
        sess_options.inter_op_num_threads = 1

        providers = self._select_providers()
        session = ort.InferenceSession(
            str(model_path), sess_options=sess_options, providers=providers
        )
        self._kokoro = Kokoro.from_session(session, str(voices_path))

        self._input_dtypes = {
            inp.name: inp.type for inp in self._kokoro.sess.get_inputs()
        }

        logger.info(
            "Kokoro TTS loaded (providers=%s, native %dHz, output %dHz)",
            self._kokoro.sess.get_providers(),
            _NATIVE_RATE,
            self.sample_rate,
        )

        self._warmup()

    @staticmethod
    def _select_providers() -> list[str]:
        available = set(ort.get_available_providers())
        preferred = [
            "CUDAExecutionProvider",
            "CoreMLExecutionProvider",
            "DmlExecutionProvider",
            "CPUExecutionProvider",
        ]
        return [p for p in preferred if p in available] or ["CPUExecutionProvider"]

    def _warmup(self) -> None:
        assert self._kokoro is not None
        voice = self._settings.tts.model or "af_heart"
        lang = "en-gb" if voice.startswith(("bf_", "bm_")) else "en-us"
        try:
            phonemes = self._kokoro.tokenizer.phonemize("Hello.", lang)
            for batch in self._kokoro._split_phonemes(phonemes):
                self._run_inference(batch, voice, 1.0)
        except Exception as e:
            logger.warning("Kokoro warmup failed: %s", e)

    @staticmethod
    def _ensure_file(filename: str, url: str) -> Path:
        """Download a file to the cache directory if not already present."""
        MODELS_DIR.mkdir(parents=True, exist_ok=True)
        path = MODELS_DIR / filename
        if not path.exists():
            logger.info("Downloading %s ...", filename)
            urllib.request.urlretrieve(url, path)
            logger.info(
                "Downloaded %s (%d MB)", filename, path.stat().st_size // 1_000_000
            )
        return path

    def _onnx_dtype(self, name: str) -> type[np.floating | np.integer]:
        """Return the numpy dtype the ONNX model expects for a given input."""
        onnx_type = self._input_dtypes.get(name, "")
        if "float" in onnx_type:
            return np.float32
        if "int64" in onnx_type:
            return np.int64
        return np.int32

    def _run_inference(
        self, phonemes: str, voice_name: str, speed: float
    ) -> np.ndarray:
        """Run ONNX inference with proper input types for quantized models."""
        tokens = self._kokoro.tokenizer.tokenize(phonemes)
        voice = self._kokoro.get_voice_style(voice_name)
        style = voice[len(tokens)]
        padded = [0, *tokens, 0]

        inputs = {
            "input_ids": np.array([padded], dtype=self._onnx_dtype("input_ids")),
            "style": np.array(style, dtype=np.float32),
            "speed": np.array([speed], dtype=np.float32),
        }
        return self._kokoro.sess.run(None, inputs)[0]

    def synthesize_iter(self, text: str) -> Iterable[bytes]:
        """Yield PCM for each phoneme batch as its inference completes.

        Kokoro splits long sentences into ~100-phoneme batches run as
        separate ORT forward passes. Yielding per batch lets the device
        start playing batch 1 while batch 2 is still inferring.
        """
        if self._kokoro is None:
            raise RuntimeError("Call load() first")

        voice = self._settings.tts.model or "af_heart"
        lang = "en-gb" if voice.startswith(("bf_", "bm_")) else "en-us"
        phonemes = self._kokoro.tokenizer.phonemize(text, lang)

        for batch in self._kokoro._split_phonemes(phonemes):
            samples = self._run_inference(batch, voice, 1.0)
            audio = (samples * 32767).clip(-32768, 32767).astype(np.int16)
            if self.sample_rate != _NATIVE_RATE:
                audio = resample(audio, _NATIVE_RATE, self.sample_rate)
            yield audio.tobytes()

    def synthesize(self, text: str) -> bytes:
        """Synthesize text to raw 16-bit PCM at self.sample_rate."""
        return b"".join(self.synthesize_iter(text))

File: internal/tts/kokoro_tts.go
package tts

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"sync"

	onnx "github.com/yalue/onnxruntime_go"

	"github.com/bryfur/ovi-voice-assistant/internal/config"
	"github.com/bryfur/ovi-voice-assistant/internal/download"
	"github.com/bryfur/ovi-voice-assistant/internal/dsp"
	"github.com/bryfur/ovi-voice-assistant/internal/ort"
)

const (
	kokoroModelURL = "https://huggingface.co/onnx-community/Kokoro-82M-v1.0-ONNX" +
		"/resolve/main/onnx/model_uint8.onnx"
	kokoroVoicesURL = "https://github.com/thewh1teagle/kokoro-onnx" +
		"/releases/download/model-files-v1.0/voices-v1.0.bin"
	kokoroNativeRate   = 24000
	kokoroMaxPhonemes  = 510
	kokoroStyleDim     = 256
	kokoroDefaultVoice = "af_heart"
)

// kokoroSymbols is the Kokoro v1.0 symbol table; index = token id.
var kokoroSymbols = func() []string {
	pad := "$"
	punctuation := ";:,.!?¡¿—…\"«»“” "
	letters := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	ipa := "ɑɐɒæɓʙβɔɕçɗɖðʤəɘɚɛɜɝɞɟʄɡɠɢʛɦɧħɥʜɨɪʝɭɬɫɮʟɱɯɰŋɳɲɴøɵɸθœɶʘɹɺɾɻʀʁɽʂʃʈʧʉʊʋⱱʌɣɤʍχʎʏʑʐʒʔʡʕʢǀǁǂǃˈˌːˑʼʴʰʱʲʷˠˤ˞↓↑→↗↘'̩'ᵻ"
	var syms []string
	syms = append(syms, pad)
	for _, r := range punctuation {
		syms = append(syms, string(r))
	}
	for _, r := range letters {
		syms = append(syms, string(r))
	}
	for _, r := range ipa {
		syms = append(syms, string(r))
	}
	return syms
}()

// KokoroVocab maps a symbol to its token id.
var KokoroVocab = func() map[rune]int64 {
	m := map[rune]int64{}
	for i, s := range kokoroSymbols {
		r := []rune(s)[0]
		if _, dup := m[r]; !dup {
			m[r] = int64(i)
		}
	}
	return m
}()

// KokoroTokenize maps phonemes to token ids, skipping unknown symbols.
func KokoroTokenize(phonemes string) []int64 {
	var ids []int64
	for _, r := range phonemes {
		if id, ok := KokoroVocab[r]; ok {
			ids = append(ids, id)
		}
	}
	return ids
}

var kokoroBatchSplit = regexp.MustCompile(`([.,!?;])`)

// KokoroSplitPhonemes splits phonemes into batches under the max length,
// preferring punctuation boundaries.
func KokoroSplitPhonemes(phonemes string) []string {
	// Split by punctuation and keep the delimiters.
	var parts []string
	last := 0
	for _, loc := range kokoroBatchSplit.FindAllStringIndex(phonemes, -1) {
		parts = append(parts, phonemes[last:loc[0]], phonemes[loc[0]:loc[1]])
		last = loc[1]
	}
	parts = append(parts, phonemes[last:])

	var batches []string
	current := ""
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		if len([]rune(current))+len([]rune(part))+1 >= kokoroMaxPhonemes {
			batches = append(batches, current)
			current = part
			continue
		}
		if strings.Contains(".,!?;", part) {
			current += part
		} else {
			if current != "" {
				current += " "
			}
			current += part
		}
	}
	if current != "" {
		batches = append(batches, current)
	}
	return batches
}

// kokoroPostProcess applies the kokoro-onnx phoneme fix-ups.
func kokoroPostProcess(phonemes, lang string) string {
	phonemes = strings.ReplaceAll(phonemes, "kəkˈoːɹoʊ", "kˈoʊkəɹoʊ")
	phonemes = strings.ReplaceAll(phonemes, "kəkˈɔːɹəʊ", "kˈəʊkəɹəʊ")
	r := strings.NewReplacer("ʲ", "j", "r", "ɹ", "x", "k", "ɬ", "l")
	phonemes = r.Replace(phonemes)
	if lang == "en-us" {
		phonemes = strings.ReplaceAll(phonemes, "nˈaɪnti", "nˈaɪndi")
	}
	var sb strings.Builder
	for _, ch := range phonemes {
		if _, ok := KokoroVocab[ch]; ok {
			sb.WriteRune(ch)
		}
	}
	return strings.TrimSpace(sb.String())
}

var (
	kokoroQuoteRe    = strings.NewReplacer("‘", "'", "’", "'", "«", "“", "»", "”", "“", "\"", "”", "\"", "(", "«", ")", "»")
	kokoroSpaceRe    = regexp.MustCompile(`[^\S \n]`)
	kokoroMultiSpace = regexp.MustCompile(`  +`)
	kokoroAbbrevs    = []struct {
		re   *regexp.Regexp
		repl string
	}{
		// Go's regexp has no lookahead, so the following context is captured
		// and re-emitted instead.
		{regexp.MustCompile(`\bD[Rr]\.( [A-Z])`), "Doctor$1"},
		{regexp.MustCompile(`\bMr\.`), "Mister"},
		{regexp.MustCompile(`\bMR\.( [A-Z])`), "Mister$1"},
		{regexp.MustCompile(`\bMs\.`), "Miss"},
		{regexp.MustCompile(`\bMS\.( [A-Z])`), "Miss$1"},
		{regexp.MustCompile(`\bMrs\.`), "Mrs"},
		{regexp.MustCompile(`\bMRS\.( [A-Z])`), "Mrs$1"},
		{regexp.MustCompile(`\betc\.( [a-z]|$)`), "etc$1"},
		{regexp.MustCompile(`(?i)\b(y)eah?\b`), "${1}e'a"},
	}
)

// KokoroNormalizeText applies light text normalisation (quotes, spacing,
// common abbreviations) before phonemization.
func KokoroNormalizeText(text string) string {
	text = kokoroQuoteRe.Replace(text)
	text = kokoroSpaceRe.ReplaceAllString(text, " ")
	text = kokoroMultiSpace.ReplaceAllString(text, " ")
	for _, a := range kokoroAbbrevs {
		text = a.re.ReplaceAllString(text, a.repl)
	}
	return strings.TrimSpace(text)
}

// KokoroTTS is text-to-speech using Kokoro 82M (ONNX int8 quantized).
type KokoroTTS struct {
	settings   *config.Settings
	sampleRate int

	mu         sync.Mutex
	session    *ort.Session
	voices     map[string]*NPYArray
	phonemizer *Phonemizer

	// ModelsDir holds cached model files.
	ModelsDir string
}

// NewKokoroTTS creates an unloaded provider targeting sampleRate.
func NewKokoroTTS(settings *config.Settings, sampleRate int) *KokoroTTS {
	if sampleRate == 0 {
		sampleRate = kokoroNativeRate
	}
	return &KokoroTTS{
		settings:   settings,
		sampleRate: sampleRate,
		ModelsDir:  filepath.Join(config.CacheDir(), "kokoro"),
	}
}

func (k *KokoroTTS) SampleRate() int  { return k.sampleRate }
func (k *KokoroTTS) SampleWidth() int { return 2 }
func (k *KokoroTTS) Channels() int    { return 1 }

func (k *KokoroTTS) voice() string {
	if v := k.settings.TTS.Model; v != "" {
		return v
	}
	return kokoroDefaultVoice
}

func (k *KokoroTTS) lang() string {
	v := k.voice()
	if strings.HasPrefix(v, "bf_") || strings.HasPrefix(v, "bm_") {
		return "en-gb"
	}
	return "en-us"
}

// Load downloads the model + voices, creates the session and warms up.
func (k *KokoroTTS) Load() error {
	if err := CheckEspeak(); err != nil {
		return err
	}
	modelPath, err := download.Ensure(filepath.Join(k.ModelsDir, "model_uint8.onnx"), kokoroModelURL)
	if err != nil {
		return err
	}
	voicesPath, err := download.Ensure(filepath.Join(k.ModelsDir, "voices-v1.0.bin"), kokoroVoicesURL)
	if err != nil {
		return err
	}
	slog.Info("Loading Kokoro TTS model")

	providers := ort.SelectProviders("auto", true)
	sess, err := ort.NewSession(modelPath, ort.SessionConfig{
		IntraOpThreads: runtime.NumCPU(),
		InterOpThreads: 1,
		Parallel:       true,
		OptExtended:    true,
		Providers:      providers,
		CoreMLCacheDir: filepath.Join(k.ModelsDir, "coreml_cache"),
	})
	if err != nil {
		return err
	}
	vdata, err := os.ReadFile(voicesPath)
	if err != nil {
		sess.Destroy()
		return err
	}
	voices, err := ParseNPZ(vdata)
	if err != nil {
		sess.Destroy()
		return fmt.Errorf("parse voices: %w", err)
	}
	if _, ok := voices[k.voice()]; !ok {
		sess.Destroy()
		return fmt.Errorf("unknown Kokoro voice %q", k.voice())
	}
	k.mu.Lock()
	k.session = sess
	k.voices = voices
	k.phonemizer = NewPhonemizer(k.lang())
	k.mu.Unlock()
	slog.Info("Kokoro TTS loaded", "providers", sess.Providers, "native_hz", kokoroNativeRate, "output_hz", k.sampleRate)
	k.warmup()
	return nil
}

func (k *KokoroTTS) warmup() {
	if err := k.SynthesizeIter("Hello.", func([]byte) error { return nil }); err != nil {
		slog.Warn("Kokoro warmup failed", "err", err)
	}
}

// Close frees the session.
func (k *KokoroTTS) Close() {
	k.mu.Lock()
	defer k.mu.Unlock()
	if k.session != nil {
		k.session.Destroy()
		k.session = nil
	}
}

// runInference runs one phoneme batch and returns 24 kHz float samples.
func (k *KokoroTTS) runInference(phonemes, voice string, speed float32) ([]float32, error) {
	tokens := KokoroTokenize(phonemes)
	if len(tokens) == 0 {
		return nil, nil
	}
	if len(tokens) > kokoroMaxPhonemes {
		tokens = tokens[:kokoroMaxPhonemes]
	}
	style := k.voices[voice]
	if style == nil || len(style.Shape) < 1 {
		return nil, fmt.Errorf("voice %q not loaded", voice)
	}
	// style array is [510, 1, 256]; pick the row for this token count.
	row := len(tokens)
	if row >= style.Shape[0] {
		row = style.Shape[0] - 1
	}
	styleVec := style.Data[row*kokoroStyleDim : (row+1)*kokoroStyleDim]

	padded := make([]int64, 0, len(tokens)+2)
	padded = append(padded, 0)
	padded = append(padded, tokens...)
	padded = append(padded, 0)

	ids, err := k.session.IntTensorFor("input_ids", []int64{1, int64(len(padded))}, padded)
	if err != nil {
		return nil, err
	}
	st, err := ort.FloatTensor([]int64{1, kokoroStyleDim}, append([]float32(nil), styleVec...))
	if err != nil {
		ids.Destroy()
		return nil, err
	}
	sp, err := ort.FloatTensor([]int64{1}, []float32{speed})
	if err != nil {
		ids.Destroy()
		st.Destroy()
		return nil, err
	}
	inputs := map[string]onnx.Value{"input_ids": ids, "style": st, "speed": sp}
	outputs, err := k.session.RunNamed(inputs)
	ort.DestroyAll([]onnx.Value{ids, st, sp})
	if err != nil {
		return nil, err
	}
	defer ort.DestroyAll(outputs)
	samples, _, err := ort.FloatData(outputs[0])
	if err != nil {
		return nil, err
	}
	return append([]float32(nil), samples...), nil
}

// SynthesizeIter yields PCM for each phoneme batch as its inference
// completes, so the device can start playing batch 1 while batch 2 is
// still inferring.
func (k *KokoroTTS) SynthesizeIter(text string, emit func(pcm []byte) error) error {
	k.mu.Lock()
	defer k.mu.Unlock()
	if k.session == nil {
		return fmt.Errorf("call Load() first")
	}
	raw, err := k.phonemizer.Phonemize(KokoroNormalizeText(text))
	if err != nil {
		return err
	}
	phonemes := kokoroPostProcess(raw, k.lang())
	for _, batch := range KokoroSplitPhonemes(phonemes) {
		samples, err := k.runInference(batch, k.voice(), 1.0)
		if err != nil {
			return err
		}
		if len(samples) == 0 {
			continue
		}
		pcm := dsp.Float32ToInt16(samples, 32767)
		if k.sampleRate != kokoroNativeRate {
			pcm = dsp.Resample(pcm, kokoroNativeRate, k.sampleRate)
		}
		if err := emit(dsp.Int16ToBytes(pcm)); err != nil {
			return err
		}
	}
	return nil
}

// Synthesize renders text to PCM at SampleRate.
func (k *KokoroTTS) Synthesize(text string) ([]byte, error) {
	return SynthesizeAll(k, text)
}
